kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/VirtualKafkaClusterReconciler.java

```diff
@@ -431,16 +419,35 @@ public List<EventSource<?, VirtualKafkaCluster>> prepareEventSources(EventSource
                 Service.class,
                 VirtualKafkaCluster.class)
                 .withName(KUBERNETES_SERVICES_EVENT_SOURCE_NAME)
-                .withPrimaryToSecondaryMapper((VirtualKafkaCluster cluster) -> cluster.getSpec().getIngresses()
-                        .stream()
-                        .map(Ingresses::getIngressRef)
-                        .flatMap(ir -> ResourcesUtil
-                                .localRefAsResourceId(cluster, new AnyLocalRefBuilder().withName(bootstrapServiceName(cluster, ir.getName())).build()).stream())
-                        .collect(Collectors.toSet()))
-                .withSecondaryToPrimaryMapper(kubenetesService -> Optional.of(kubenetesService)
-                        .flatMap(service -> extractOwnerRefFromKubernetesService(service, VIRTUAL_KAFKA_CLUSTER_KIND))
-                        .map(ownerRef -> new ResourceID(ownerRef.getName(), kubenetesService.getMetadata().getNamespace()))
-                        .map(Set::of).orElse(Set.of()))
+                .withPrimaryToSecondaryMapper((VirtualKafkaCluster cluster) -> {
+                    Stream<ResourceID> clusterIpServices = cluster.getSpec().getIngresses()
+                            .stream()
+                            .map(Ingresses::getIngressRef)
+                            .flatMap(ir -> ResourcesUtil
+                                    .localRefAsResourceId(cluster, new AnyLocalRefBuilder().withName(bootstrapServiceName(cluster, ir.getName())).build()).stream());
+                    Stream<ResourceID> loadbalancerService = ResourcesUtil.localRefAsResourceId(cluster,
+                            new AnyLocalRefBuilder().withName(cluster.getSpec().getProxyRef().getName() + "-sni").build()).stream();
+                    return Stream.concat(clusterIpServices, loadbalancerService).collect(Collectors.toSet());
+                })
+                .withSecondaryToPrimaryMapper(kubernetesService -> {
+                    Optional<OwnerReference> ownerReference = extractOwnerRefFromKubernetesService(kubernetesService, VIRTUAL_KAFKA_CLUSTER_KIND);
+                    Optional<OwnerReference> ownerReferenceProxy = extractOwnerRefFromKubernetesService(kubernetesService, KAFKA_PROXY_KIND);
+                    if (ownerReference.isPresent()) {
+                        return ownerReference
+                                .map(ownerRef -> new ResourceID(ownerRef.getName(), kubernetesService.getMetadata().getNamespace()))
+                                .map(Set::of).orElse(Set.of());
+                    }
+                    else if (ownerReferenceProxy.isPresent()) {
+                        List<VirtualKafkaCluster> items = context.getClient().resources(VirtualKafkaCluster.class).inNamespace(namespace(kubernetesService)).list()
+                                .getItems();
+                        return items.stream().filter(cluster -> cluster.getSpec().getProxyRef().getName().equals(ownerReferenceProxy.get().getName()))
```

Comment:
Perhaps this ought to be a generic method on `ResourceUtils`?

Reply:
have shuffled it to ResourceUtils, pretty oddball case though, finding all resources of a particular type that reference a particular OwnerReference. Maybe it would be better to resolve the OwnerReference into the actual owner Resource, then we could use the existing utils to obtain the referrers.